core/trino-main/src/main/java/io/trino/sql/gen/DereferenceCodeGenerator.java

```diff
@@ -44,7 +44,7 @@ public DereferenceCodeGenerator(SpecialForm specialForm)
         returnType = specialForm.getType();
         checkArgument(specialForm.getArguments().size() == 2);
         base = specialForm.getArguments().get(0);
-        index = (int) ((ConstantExpression) specialForm.getArguments().get(1)).getValue();
+        index = toIntValue(((ConstantExpression) specialForm.getArguments().get(1)).getValue());
```

Comment:
Why would dereference index be polymorhic?
i would hope it's always int, or always long; always same type.

Reply:
It seems it's not but I cannot track the origin of this problem.